cxx/distributions/bigram.hh

```diff
@@ -18,11 +18,18 @@ class Bigram : public Distribution<std::string> {
 
  public:
   double alpha = 1;  // hyperparameter for all transition distributions.
-  size_t num_chars = '~' - ' ' + 1;  // printable ASCII without DEL.
-  mutable std::vector<DirichletCategorical> transition_dists;
   size_t max_length = 0;  // 0 means no maximum length
+  char min_char;  // Character with smallest ASCII value.
+  char max_char;  // Character with largest ASCII value.
+  size_t num_chars;
+  mutable std::vector<DirichletCategorical> transition_dists;
 
-  Bigram(size_t _max_length = 80): max_length(_max_length) {
+  Bigram(size_t _max_length = 80,
+         char _min_char = ' ',
```

Comment:
Is there a reason these need to be constructor arguments / modifiable? Should/can we keep these as  macro constants / static const members?

Reply:
Yeah, so the reason I changed them to be constructor arguments is that they are used in constructor (to calculate num_chars, which in turn is used to make the transition_dist tables).

I wanted the subclass to change them, but also be able to call the parent's constructor to do the necessary work.